### PR TITLE
When backend has uses_r_object capability, encode correctly (e.g. handoff)

### PR DESCRIPTION
**Pull request by slfritchie**
Required bugfix for 1.4.0 release, sorry I missed it earlier.

### CODE REVIEW

File: src/riak_object.erl [hunk 1]
```diff
@@ -605,8 +605,11 @@ to_binary_version(v0, _, _, <<131,_/binary>>=Bin) ->
     Bin;
 to_binary_version(v1, _, _, <<?MAGIC:8/integer, 1:8/integer, _/binary>>=Bin) ->
     Bin;
-to_binary_version(Vsn, B, K, Bin) ->
-    to_binary(Vsn, from_binary(B, K, Bin)).
+to_binary_version(Vsn, B, K, Bin) when is_binary(Bin) ->
+    to_binary(Vsn, from_binary(B, K, Bin));
+%% r_object record definition isn't available, so hack the guard.
```
Comment (jrwest) [inline src/riak_object.erl]: not sure i follow, why can't this just match on the r_object defined in this module? im probably missing something...

Comment (slfritchie) [inline src/riak_object.erl]: Heh, cut-and-paste error from another module, sorry, good catch.  Will push a fix in a moment.

File: src/riak_object.erl [hunk 2]
```diff
@@ -605,8 +605,10 @@ to_binary_version(v0, _, _, <<131,_/binary>>=Bin) ->
     Bin;
 to_binary_version(v1, _, _, <<?MAGIC:8/integer, 1:8/integer, _/binary>>=Bin) ->
     Bin;
-to_binary_version(Vsn, B, K, Bin) ->
-    to_binary(Vsn, from_binary(B, K, Bin)).
+to_binary_version(Vsn, B, K, Bin) when is_binary(Bin) ->
+    to_binary(Vsn, from_binary(B, K, Bin));
+to_binary_version(Vsn, _B, _K, O) when is_record(O, r_object) ->
```
Comment (jrwest) [inline src/riak_object.erl]: sorry a little more nitpicking. doesn't most of the riak code use the `#r_object{}` style [1]

[1] https://github.com/basho/riak_kv/blob/master/src/riak_object.erl#L114

**Comment by jrwest** [comment]
@slfritchie to make sure I understand correctly: this change is necessary because backends using `uses_r_object` pass an `r_object` instead of the binary to the fold which would crash handoff when that `r_object` was passed into `to_binary_version`. Do you have a quick way I can test/reproduce?

In hindsight i wonder if `to_binary_version` is even necessary or if I should have just made it part of `to_binary`. Meh, too much of a change this late in the game.

**Comment by slfritchie** [comment]
Use a file system that has the new capability, `uses_r_object`.  (Sorry, the originaly commit tried using backticks for GH markdown, but my terminal shell replaced them for me, oops.)  At the moment, there's only one: fs2.  Otherwise, hack on the yessir backend to support `uses_r_object`.

Evidence in the (lab) wild:

```
2013-06-07 16:03:43.962 [error] <0.25310.7> gen_server <0.25310.7> terminated with reason: no function clause matching riak_object:from_binary(<<48,98,58,228,222,232,92,94,66,158,249,217,95,172,49,136,69,118,125>>, <<23,163,204,214,164,168,79,24,166,171,164,136,221,38,197,117,0,0,0,39>>, {r_object,<<48,98,58,228,222,232,92,94,66,158,249,217,95,172,49,136,69,118,125>>,<<23,163,204,214,...>>,...}) line 618
```

**Comment by slfritchie** [comment]
ping.  :-)

**Comment by jrwest** [comment]
@slfritchie meant to ping it this morning my bad. Went searching for new fs backend to test with but couldn't find a branch. Do you have it available? Otherwise this change is pretty straight forward. I'll run a few riak tests when I get home and give it the official +1

**Comment by jrwest** [comment]
checked `verify_riak_object_format` still passes (note the test itself needs to be updated for 1.4 but can be made to pass easily for review). Also, setup a devrel and made sure normal handoff operations are not affected. Did not have a chance to test w/ the updated fs2 backend since I dont have a copy but I think its sufficient that this issue affects only backends using the new capability and @slfritchie reports the issue is gone w/ this PR. ~~+1.~~ (removed +1 since more changes came)

**Comment by jrwest** [comment]
@slfritchie imo the last commit should be a seperate PR. The change seems to be related to a different bug introduced by different patch.

**Comment by slfritchie** [comment]
Don't you know that the code freeze is happening any second now?

**Comment by jrwest** [comment]
well technically it happened already....lets call these bug fixes found during testing of 1.4?

**Comment by jrwest** [comment]
Also, I'm not sure a better fix would be to update `riak_core:get_primary_apl/3` to take either the service or the up nodes. The chashbin changes should have made it easy to no longer pass in the ring with minimal interface change (e.g. call the 3-arity version w/o the ring). This is why this should be in a seperate PR, so it can be discussed w/o ruining the discussion history of the original changes in this one.

**Comment by jrwest** [comment]
tested again w/ the updated fs2 backend [1]. Confirmed that there were no errors when doing handoff w/ data. However, I was unable to reproduce the `riak_object:from_binary` error in the comment above in this PR by removing the change. I was able to produce a badmatch on `riak_object:to_binary_version` during handoff by removing that change. @slfritchie any details on how I can reproduce the `from_binary` error (I thought handoff would do it)?

NOTE: all the above was tested w/o f17cd6e

My config in `riak_kv` section of app.config for fs2 backend:

```
{add_paths, ["/home/jordan/sandbox/basho/review/slf-backend-robj-cap/riak_cs/ebin"]},
{storage_backend, riak_kv_fs2_backend},
{data_root, "./data/fs2"}, %% fs2 data root
{block_size, 1024}, %% fs2 block size, picked a random power of 2
```

[1] https://github.com/basho/riak_cs/blob/gh555-fs2-stability/src/riak_kv_fs2_backend.erl

**Comment by slfritchie** [comment]
Jordan, I saw handoff do it on a Riak cluster with fs2 underneath and Riak CS the sole app talking to Riak. {shrug}

**Comment by jrwest** [comment]
@slfritchie ok. i'll have to take your word for it since I am still unable to reproduce w/ just riak + fs2. looking at the other change in this PR re: n_val now

**Comment by buddhisthead** [comment]
+1 based on https://github.com/basho/riak_test/pull/304, which used to exhibit the error: insufficient_vnodes before this fix. With it, the test passes.

**Comment by jrwest** [comment]
with @buddhisthead's verification + my testing I'm +1 on this as well.

**Comment by slfritchie** [comment]
https://github.com/basho/riak_core/issues/337 created for lingering review comment.

**Pull request merged**